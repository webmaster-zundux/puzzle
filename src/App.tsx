import "@fontsource-variable/open-sans";
import type { FC } from "react";
import { memo } from "react";
import "./styles.css";

import { PuzzleDataProvider } from "./components/PuzzleDataProvider";
import { PuzzlePage } from "./pages/PuzzlePage";

export const App: FC = memo(function App() {
  return (
    <PuzzleDataProvider>
      <PuzzlePage />
    </PuzzleDataProvider>
  );
});
