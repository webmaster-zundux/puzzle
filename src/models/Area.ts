import { Point } from "./Point";

export class Area {
  constructor(
    public p0: Point,
    public p1: Point,
    public p2: Point,
    public p3: Point,
  ) {}

  public static create({
    p0 = Point.getZeroPoint(),
    p1 = Point.getZeroPoint(),
    p2 = Point.getZeroPoint(),
    p3 = Point.getZeroPoint(),
  }: {
    p0: Point;
    p1: Point;
    p2: Point;
    p3: Point;
  }) {
    return new Area(p0, p1, p2, p3);
  }

  public static isPointWithinArea(point: Point | { x: number; y: number }, area: Area): boolean {
    const { p0, p2 } = area;
    const { x: px, y: py } = point;

    if (px < p0.x || px > p2.x) {
      return false;
    }

    if (py < p0.y || py > p2.y) {
      return false;
    }

    return true;
  }

  public static isPointWithinAreaByCornerPoints(
    point: Point | { x: number; y: number },
    areaPointP0: Point | { x: number; y: number },
    areaPointP2: Point | { x: number; y: number },
  ): boolean {
    const tx = point.x;
    const ty = point.y;

    const p0 = areaPointP0;
    const p2 = areaPointP2;

    if (tx < p0.x || tx > p2.x) {
      return false;
    }

    if (ty < p0.y || ty > p2.y) {
      return false;
    }

    return true;
  }

  public static doTwoAreasIntersect(areaA: Area, areaB: Area): boolean {
    const aP0 = areaA.p0;
    const aP2 = areaA.p2;

    const bP0 = areaB.p0;
    const bP2 = areaB.p2;

    return Area.doTwoAreasIntersectByCornerPoints(aP0, aP2, bP0, bP2);
  }

  public static doTwoAreasIntersectByCornerPoints(
    areaAPointP0: Point | { x: number; y: number },
    areaAPointP2: Point | { x: number; y: number },
    areaBPointP0: Point | { x: number; y: number },
    areaBPointP2: Point | { x: number; y: number },
  ): boolean {
    const aP0 = areaAPointP0;
    const aP2 = areaAPointP2;

    const bP0 = areaBPointP0;
    const bP2 = areaBPointP2;

    let intersectionOnXAxis = true;
    let intersectionOnYAxis = true;

    if (aP2.x < bP0.x || aP0.x > bP2.x) {
      intersectionOnXAxis = false;
    }

    if (aP2.y < bP0.y || aP0.y > bP2.y) {
      intersectionOnYAxis = false;
    }

    return intersectionOnXAxis && intersectionOnYAxis;
  }

  public static getDefaultArea() {
    return new Area(new Point(0, 0), new Point(1, 0), new Point(1, 1), new Point(0, 1));
  }

  toString() {
    return `Area {p0: ${this.p0}, p1: ${this.p1}, p2: ${this.p2}, p3: ${this.p3}}`;
  }
}
