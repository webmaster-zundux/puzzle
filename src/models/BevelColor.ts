export const BEVEL_EFFECT_HIGHLIGHT = "highlight" as const;
export const BEVEL_EFFECT_SHADOW = "shadow" as const;

export type BEVEL_EFFECT_TYPE = typeof BEVEL_EFFECT_SHADOW | typeof BEVEL_EFFECT_HIGHLIGHT;
