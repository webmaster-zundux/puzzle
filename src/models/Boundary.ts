export class Boundary {
  constructor(
    public x: number,
    public y: number,
    public width: number,
    public height: number,
  ) {}

  public static getIntersectionBoundaryOfTwoBoundaries(a: Boundary, b: Boundary): Boundary | undefined {
    const x1 = a.x > b.x ? a.x : b.x;
    const y1 = a.y > b.y ? a.y : b.y;

    const aP2X = a.x + a.width;
    const aP2Y = a.y + a.height;

    const bP2X = b.x + b.width;
    const bP2Y = b.y + b.height;

    const x2 = aP2X < bP2X ? aP2X : bP2X;
    const y2 = aP2Y < bP2Y ? aP2Y : bP2Y;

    const width = x2 - x1;
    const height = y2 - y1;

    // checking if they are intersect
    if (width <= 0 || height <= 0) {
      return undefined;
    }

    return {
      x: x1,
      y: y1,
      width,
      height,
    };
  }
}
