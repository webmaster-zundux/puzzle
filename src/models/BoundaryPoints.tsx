import type { Point } from "./Point";

export type BoundaryPoints = {
  tl: Point;
  br: Point;
};
