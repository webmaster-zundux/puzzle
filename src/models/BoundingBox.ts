export class BoundingBox {
  constructor(
    public x: number = 0,
    public y: number = 0,
    public width: number = 1,
    public height: number = 0,
  ) {}

  public static create({ x, y, width, height }: { x: number; y: number; width: number; height: number }) {
    return new BoundingBox(x, y, width, height);
  }
}
