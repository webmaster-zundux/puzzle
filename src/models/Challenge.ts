import type { Id } from "../core/puzzle/BaseEntity";
import { uuid } from "../utils/uuid";

export type ChallengeId = Id;

const generateChallengeIdFunction = () => {
  const newUUID = uuid();
  return newUUID;
};

const EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT =
  import.meta.env.VITE_TESTING_CHALLENGE_ID_GENERATION_EXPOSE_TO_WINDOW_OBJECT === "1";
if (EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT) {
  // @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
  if (typeof window?.generateChallengeId !== "function") {
    Object.defineProperty(window, "generateChallengeId", {
      value: generateChallengeIdFunction,
      writable: false,
      configurable: false,
    });
  }
}

// @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
export const generateChallengeId = window?.generateChallengeId ?? generateChallengeIdFunction;
