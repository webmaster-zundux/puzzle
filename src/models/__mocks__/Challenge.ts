export const generateChallengeId = () => {
  const newUUID = "364bf6d0-6686-4892-be7d-618ac2fcdcc7";
  return newUUID;
};
