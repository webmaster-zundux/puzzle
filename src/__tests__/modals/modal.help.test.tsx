import { screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { App } from "../../App";
import * as Challenge from "../../models/Challenge";
import { renderIntoDocumentBody } from "../__tests-utils__/renderIntoDocumentBody";
import { openHelpDialogByClickOnHelpButtonInSidebar } from "./modal-helpers";

describe(`"help" dialog`, () => {
  describe(`changes it's visibility state`, () => {
    it(`opens when clicking on "help" button in the sidebar`, async () => {
      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();

      await openHelpDialogByClickOnHelpButtonInSidebar({ isShortCheck: false });

      expect(renderWithoutDialog).toMatchDiffSnapshot(
        asFragment(),
        { aAnnotation: "without dialog", bAnnotation: "with dialog" },
        `"help" dialog`,
      );
    });

    it("closes when clicking on the dialog backdrop", async () => {
      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();
      const dialog = await openHelpDialogByClickOnHelpButtonInSidebar();
      const backdrop = within(dialog).getByLabelText(/modal-backdrop/i);
      expect(backdrop).toBeInTheDocument();

      await userEvent.click(backdrop!);

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without dialog`);
    });

    it('closes when pressing on "esc" key', async () => {
      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();
      await openHelpDialogByClickOnHelpButtonInSidebar();

      await userEvent.keyboard("{Escape}");

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without dialog`);
    });

    it('closes when clicking on "close" button', async () => {
      vi.spyOn(Challenge, "generateChallengeId")
        .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
        .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");

      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();
      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      const initialChallengeSeed = screen.getByRole("note", {
        name: /challenge seed:/i,
      }).textContent;
      expect(initialChallengeSeed).toBe("b47663e4-95ab-4472-9344-0f165a6a7ff9");
      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      const dialog = await openHelpDialogByClickOnHelpButtonInSidebar();

      await userEvent.click(within(dialog).getByRole("button", { name: /close/i }));

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without dialog`);
      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      const challengeSeedAfterCanceledRestartOfPuzzle = screen.getByRole("note", {
        name: /challenge seed:/i,
      }).textContent;
      expect(challengeSeedAfterCanceledRestartOfPuzzle).toBe(initialChallengeSeed);
    });
  });
});
