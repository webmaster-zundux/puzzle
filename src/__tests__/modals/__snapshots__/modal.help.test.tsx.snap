// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`src/__tests__/modals/modal.help.test.tsx > "help" dialog > changes it's visibility state > closes when clicking on "close" button: without dialog 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`src/__tests__/modals/modal.help.test.tsx > "help" dialog > changes it's visibility state > closes when clicking on the dialog backdrop: without dialog 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`src/__tests__/modals/modal.help.test.tsx > "help" dialog > changes it's visibility state > closes when pressing on "esc" key: without dialog 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`src/__tests__/modals/modal.help.test.tsx > "help" dialog > changes it's visibility state > opens when clicking on "help" button in the sidebar: "help" dialog 1`] = `
"Snapshot Diff:
- without dialog
+ with dialog

@@ -453,6 +453,85 @@
            </div>
          </div>
        </aside>
      </main>
    </div>
+   <div>
+     <div
+       aria-label="Help"
+       aria-modal="true"
+       class="_Dialog_9b989f _DialogAsDiv_9b989f"
+       role="dialog"
+     >
+       <div
+         class="_DialogBodyContainer_9b989f"
+       >
+         <div
+           aria-hidden="true"
+           aria-label="modal-backdrop"
+           class="_Backdrop_9b989f"
+         />
+         <div
+           class="_DialogBody_9b989f _WithCardForContent_9b989f"
+         >
+           <div
+             class="_CloseIcon_bea9a0"
+           >
+             <button
+               aria-label="close"
+               class="_ActionButton_b70b4c _NoBackground_b70b4c _NoBorder_b70b4c"
+               type="button"
+             >
+               <div
+                 class="_Icon_14488b _DoubleSize_14488b"
+                 data-src="/assets/fa-icons/xmark.svg"
+               >
+                 inline-svg-icon-mock
+               </div>
+             </button>
+           </div>
+           <div
+             class="_DialogContent_9b989f"
+           >
+             <div
+               class="_Content_a024cf"
+             >
+               <h3
+                 class="_Title_a024cf"
+               >
+                 Help
+               </h3>
+               <div
+                 class="_MainContent_a024cf"
+               >
+                 <div
+                   class="_InterfaceScreenshotContainer_178972"
+                 >
+                   <div
+                     class="_InterfaceScreenshot_178972"
+                   >
+                     interface screenshots
+                   </div>
+                   <div
+                     class="_InterfaceScreenshotLegendLabel_178972"
+                   >
+                     legend label for screenshot
+                   </div>
+                 </div>
+                 <div
+                   class="_AnimatedPuzzleSolvingGifContainer_178972"
+                 >
+                   <img
+                     alt="animated puzzle solving gif"
+                     class="_AnimatedPuzzleSolvingGif_178972"
+                     src="#animated-puzzle-solving-gif"
+                     title="animated puzzle solving gif"
+                   />
+                 </div>
+               </div>
+             </div>
+           </div>
+         </div>
+       </div>
+     </div>
+   </div>
  </DocumentFragment>"
`;
